src/Features/Core/Portable/FindUsages/DefinitionItemFactory.cs

```diff
@@ -151,75 +116,108 @@ internal static class DefinitionItemExtensions
                     tags, displayParts,
-                    DefinitionItem.GetOriginationParts(definition),
+                    nameDisplayParts,
+                    metadataLocations,
                     properties, displayIfNoReferences);
             }
 
             var displayableProperties = AbstractReferenceFinder.GetAdditionalFindUsagesProperties(definition);
 
             return DefinitionItem.Create(
-                tags, displayParts, sourceLocations, classifiedSpans,
+                tags, displayParts, sourceLocations, classifiedSpans, metadataLocations,
                 nameDisplayParts, properties, displayableProperties, displayIfNoReferences);
         }
 
-        private static ImmutableArray<DocumentSpan> TryGetSourceLocations(ISymbol definition, Solution solution, ImmutableArray<Location> locations, bool includeHiddenLocations)
+        internal static ImmutableArray<AssemblyLocation> GetMetadataLocations(ISymbol definition, Solution solution, out ProjectId? originatingProjectId)
         {
-            // If it's a namespace, don't create any normal location.  Namespaces
-            // come from many different sources, but we'll only show a single 
-            // root definition node for it.  That node won't be navigable.
-            if (definition.Kind == SymbolKind.Namespace)
+            originatingProjectId = null;
+
+            if (!definition.Locations.Any(static location => location.MetadataModule != null))
             {
                 return [];
             }
 
-            // If it's a namespace, don't create any normal location.  Namespaces
-            // come from many different sources, but we'll only show a single 
-            // root definition node for it.  That node won't be navigable.
-            using var sourceLocations = TemporaryArray<DocumentSpan>.Empty;
+            var assembly = definition as IAssemblySymbol ?? definition.ContainingAssembly;
+            if (assembly != null)
+            {
+                // symbol is defined within a single metadata assembly:
+                var info = solution.CompilationState.GetOriginatingProjectInfo(assembly);
+                Contract.ThrowIfNull(info);
+                Contract.ThrowIfNull(info.ReferencedThrough);
 
-            foreach (var location in locations)
+                originatingProjectId = info.ProjectId;
+                return [new AssemblyLocation(assembly.Identity.Name, assembly.Identity.Version, info.ReferencedThrough.Value.FilePath)];
+            }
+
+            if (definition is INamespaceSymbol namespaceSymbol)
             {
-                if (location.IsInMetadata)
-                {
-                    return default;
-                }
+                using var metadataLocations = TemporaryArray<AssemblyLocation>.Empty;
+
+                // only shared namespace symbols don't have containing assembly:
+                Contract.ThrowIfTrue(namespaceSymbol.ConstituentNamespaces.IsEmpty);
```

Comment:
how are you sure this assert is trye?  i't's not reaslly clear to me why we must have constituents.  is is some invariant of a namespace that has a metadatalocation?

Reply:
The doc on [`ISymbol.ContainingAssembly`](https://github.com/dotnet/roslyn/blob/e4b57d382701f0c498ec0629f57352e1ff6c03a2/src/Compilers/Core/Portable/Symbols/ISymbol.cs#L73-L77) says it's null for symbols shared across assemblies.

The previous condition handled cases where `ContainingAssembly` is non-null.
It follows that the namespace symbol has constituents, otherwise it would have non-null `ContainingAssembly`

Reply:
wfm.  consider doc'ing.